lib/cext/preprocess.rb

```diff
@@ -145,6 +160,201 @@
+      },
+      {
+        match: /wrapper->rows = rb_ary_new\(\);/,
+        replacement: 'wrapper->rows = rb_tr_managed_from_handle(rb_ary_new());'
+      },
+      {
+        match: /rb_yield\(rb_ary_entry\(wrapper->rows, i\)\);/,
+        replacement: 'rb_yield(rb_ary_entry(rb_tr_managed_from_handle(wrapper->rows), i));'
+      },
+      {
+        match: /RARRAY_LEN\(wrapper->rows\)/,
+        replacement: 'RARRAY_LEN(rb_tr_managed_from_handle(wrapper->rows))'
+      },
+      {
+        match: /rb_ary_entry\(wrapper->rows,/,
+        replacement: 'rb_ary_entry(rb_tr_managed_from_handle(wrapper->rows),'
+      },
+      {
+        match: /rb_ary_store\(wrapper->rows,/,
+        replacement: 'rb_ary_store(rb_tr_managed_from_handle(wrapper->rows),'
+      },
+      {
+        match: /return wrapper->rows;/,
+        replacement: 'return rb_tr_managed_from_handle(wrapper->rows);'
+      },
+      {
+        match: /wrapper->rows = rb_ary_new2\(wrapper->numberOfRows\);/,
+        replacement: 'wrapper->rows = rb_tr_handle_for_managed(rb_ary_new2(wrapper->numberOfRows));'
+      },
+      {
+        match: /args\.db_timezone = db_timezone;/,
+        replacement: 'args.db_timezone = rb_tr_handle_for_managed_leaking(db_timezone);'
+      },
+      {
+        match: /args\.app_timezone = app_timezone;/,
+        replacement: 'args.app_timezone = rb_tr_handle_for_managed_leaking(app_timezone);'
+      },
+      {
+        match: /args\.block_given = block;/,
+        replacement: 'args.block_given = rb_tr_handle_for_managed_leaking(block);'
+      },
+      {
+        match: /wrapper->fields = Qnil/,
+        replacement: 'wrapper->fields = rb_tr_handle_for_managed(Qnil)'
+      },
+      {
+        match: /wrapper->rows = Qnil/,
+        replacement: 'wrapper->rows = rb_tr_handle_for_managed(Qnil)'
+      },
+      {
+        match: /wrapper->encoding = encoding/,
+        replacement: 'wrapper->encoding = rb_tr_handle_for_managed(encoding)'
+      },
+      {
+        match: /wrapper->client = client/,
+        replacement: 'wrapper->client = rb_tr_handle_for_managed(client)'
+      },
+      {
+        match: /wrapper->statement = statement/,
+        replacement: 'wrapper->statement = rb_tr_handle_for_managed_leaking(statement)'
+      }
+    ]
+  },
+  'statement.c' => {
+    gem: 'mysql2',
+    patches: [
+      {
+        match: /rb_to_encoding\(wrapper->encoding\)/,
+        replacement: 'rb_to_encoding(rb_tr_managed_from_handle(wrapper->encoding))'
+      },
+      {
+        match: /wrapper->active_thread = Qnil/,
+        replacement: 'wrapper->active_thread = rb_tr_handle_for_managed(Qnil)'
+      },
+      {
+        match: /rb_mysql_result_to_obj\(stmt_wrapper->client, wrapper->encoding, current, metadata, self\)/,
+        replacement: 'rb_mysql_result_to_obj(stmt_wrapper->client, rb_tr_managed_from_handle(wrapper->encoding), current, metadata, self)'
+      }
+    ]
```

Comment:
That's a long list! And basically just to handle/de-handle a couple fields of 2 `struct`s.
@grimmerm It would be very nice if somehow we could annotate those fields (in the C code) so that Sulong handle/de-handle them automatically or let us install a hook to do it.
The lifetime also seems very simple here for most fields, handles can be released whenever the struct itself is free'd.